// grunt-webify/tasks/webify.js
var htmlparser = require("html-parser");
var ejs = require("ejs");
var path = require("path");
var sass = require("node-sass");
var uu = require("underscore");
var us = require("underscore.string");
var browserify = require("browserify-string");
var async = require("async");
var fs = require("fs");
var es6ify = require("es6ify");
var Compiler = require("../util/es6-module-transpiler.js").Compiler;

var buildHTML = function(file, options, cb) {
    var currentEl = null;
    var output = "";

    // Concatenations
    var concatenatedScripts = "";
    var concatenatedSass = "";
    var concatenatedStyles = "";

    htmlparser.parse(file, {
	openElement: function(name) {
	    currentEl = name;
	    if(name !== "script" && currentEl !== "style") {
		output += "<" + name;
	    }
	},

	attribute: function(key, value) {
	    if(currentEl === "script") {
		if(key === "src") {
		    concatenatedScripts += fs.readFileSync(path.resolve(options.path, value));
		}
	    } else if(currentEl === "style") {
		if(key === "src") {
		    if(us.endsWith(value, ".scss")) {
			concatenatedSass += fs.readFileSync(path.resolve(options.path, value));
		    } else {
			concatenatedStyles += fs.readFileSync(path.resolve(options.path, value));
		    }
		}
	    } else {
		output += (" " + key + '="' + value + '"');
	    }
	},

	closeOpenedElement: function(name, token) {
	    if(currentEl !== "script" && currentEl !== "style") {
		output += token;
	    }
	},

	closeElement: function(name) {
	    if(name !== "script" && name !== "style") {
		output += ("</" + name + ">");
	    }
	    currentEl = null;
	},

	text: function(value) {
	    if(currentEl === "script") {
		concatenatedScripts += value;
	    } else if(currentEl === "style") {
		concatenatedStyles += value;
	    } else {
		output += value;
	    }
	}
    });

    // Insert JavaScripts
    modulify(concatenatedScripts, function(err, src) {
	if(err) {
	    throw err;
	}
	var x = browserify(src);
	if(options.esnext) {
	    x = x.transform(es6ify);
	}
	x.bundle(function(err, src) {
	    if(err) {
		// Do not allow another cb
		return cb(err);
	    }

	    output += "<script type=\"text/javascript\">" + src + "</script>";

	    sass.render(uu.extend(options.sass, {
		data: concatenatedSass,
		success: function(css) {
		    cb(null, output + "<style type=\"text/css\">" + css +concatenatedStyles + "</style>");
		}
	    }));
	});
    });
};

var modulify = function(src, cb) {
    cb(null, new Compiler(src).toCJS());
};

var buildJS = function(src, options, cb) {
    modulify(src, function(err, output) {
	if(err) {
	    return cb(err);
	}
	cb(null, output);
    });
};

var compileToFunction = function(src, options, cb) {
    cb(null, "module.exports = exports = " + ejs.compile(src, {
	client: true,
	filename: options.path
    }));
};

var buildEJS = function(src, options, cb) {
    if(src.indexOf("<head>") !== -1) {
	src = src.split(/(<head>|<\/head>)/, 5);
	buildHTML(src[2], options, function(err, output) {
	    if(err) {
		return cb(err);
	    }
	    compileToFunction(src[0] + src[1] + output  + src[3] + src[4], options,  cb);
	});
    } else {
	compileToFunction(src, options, cb);
    }
};

var buildAny = function(options) {
    return function(file, cb) {
	options = uu.clone(options);
	options.path = path.dirname(path.resolve(file));
	fs.readFile(file, function(err, src) {
	    if(err) {
		return cb(err);
	    } else {
		src = src.toString();
		if(us.endsWith(file, ".js")) {
		    buildJS(src, options, cb);
		} else {
		    buildEJS(src, options, cb);
		}
	    }
	});
    };
};

var build = function(options) {
    return function(file, cb) {
	async.map(file.src, buildAny(options), function(err, files) {
	    if(err) return cb(err);
	    fs.writeFileSync(file.dest, options.banner + " " + files.join(""));
	    console.log("File " + file.dest + " created");
	    cb(null);
	});
    };
};

module.exports = function(grunt) {
    grunt.registerMultiTask("webify", function() {
	var done = this.async();
	var options = this.options({
	    esnext: false,
	    banner: "",
	    sass: {
		includePaths: []
	    }
	});
	if(!options.sass.includePaths) options.sass.includePaths = [];
	options.sass.includePaths.push(path.resolve(__dirname, "..", "foundation/scss"));
	async.each(this.files, build(options), function(err) {
	    if(err) grunt.log.error(err);
	    done();
	});
    });
};
